### ISSUE TITLE
精调模型不成功

### ISSUE THREAD

**Issue by ibmxiang** [ymcui/Chinese-LLaMA-Alpaca-2#182]
### 提交前必须检查以下项目

- [X] 请确保使用的是仓库最新代码（git pull），一些问题已被解决和修复。
- [X] 我已阅读[项目文档](https://github.com/ymcui/Chinese-LLaMA-Alpaca-2/wiki)和[FAQ章节](https://github.com/ymcui/Chinese-LLaMA-Alpaca-2/wiki/常见问题)并且已在Issue中对问题进行了搜索，没有找到相似问题和解决方案。
- [X] 第三方插件问题：例如[llama.cpp](https://github.com/ggerganov/llama.cpp)、[LangChain](https://github.com/hwchase17/langchain)、[text-generation-webui](https://github.com/oobabooga/text-generation-webui)等，同时建议到对应的项目中查找解决方案。

### 问题类型

模型训练与精调

### 基础模型

Alpaca-2-7B

### 操作系统

Linux

### 详细描述问题

对模型进行精调-在完整模型Chinese-LLaMA-Alpaca-2上进行微调，
模型微调运行成功，但是推理的时候代码还是回答的以前问题。请问是哪里出现了问题？我尝试过Chinese-LLaMA-Alpaca-2，以及Chinese-LLaMA-Alpaca
数据集为几十条身份信息数据。
如：  {
    "instruction": "你身份是什么？",
    "input": "",
    "output": "我叫ai助手。我的主要目标是协助用户解决问题和满足他们的需求。"
  },
而我推理代码如下：
`python3 scripts/inference/inference_hf.py \
    --base_model /root/models/chinese-alpaca-2-7b \
    --lora_model /root/Chinese-LLaMA-Alpaca-2/models/sft_lora_model \
    --with_prompt \
    --interactive`
```
微调代码如下
########参数部分########
lr=1e-4
lora_rank=64
lora_alpha=128
lora_trainable="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj"
lora_dropout=0.05

pretrained_model=/root/models/chinese-alpaca-2-7b
chinese_tokenizer_path=/root/models/chinese-alpaca-2-7b
dataset_dir=/root/Chinese-LLaMA-Alpaca-2/datas/sft
per_device_train_batch_size=1
per_device_eval_batch_size=1
gradient_accumulation_steps=1
output_dir=/root/Chinese-LLaMA-Alpaca-2/models
validation_file=/root/Chinese-LLaMA-Alpaca-2/datas/sft/self_cognition.json
max_seq_length=128

deepspeed_config_file=ds_zero2_no_offload.json

########启动命令########
torchrun --nnodes 1 --nproc_per_node 2 run_clm_sft_with_peft.py \
    --deepspeed ${deepspeed_config_file} \
    --model_name_or_path ${pretrained_model} \
    --tokenizer_name_or_path ${chinese_tokenizer_path} \
    --dataset_dir ${dataset_dir} \
    --per_device_train_batch_size ${per_device_train_batch_size} \
    --per_device_eval_batch_size ${per_device_eval_batch_size} \
    --do_train \
    --do_eval \
    --seed $RANDOM \
    --fp16 \
    --num_train_epochs 2 \
    --lr_scheduler_type cosine \
    --learning_rate ${lr} \
    --warmup_ratio 0.03 \
    --weight_decay 0 \
    --logging_strategy steps \
    --logging_steps 10 \
    --save_strategy steps \
    --save_total_limit 3 \
    --evaluation_strategy steps \
    --eval_steps 250 \
    --save_steps 500 \
    --gradient_accumulation_steps ${gradient_accumulation_steps} \
    --preprocessing_num_workers 8 \
    --max_seq_length ${max_seq_length} \
    --output_dir ${output_dir} \
    --overwrite_output_dir \
    --ddp_timeout 30000 \
    --logging_first_step True \
    --lora_rank ${lora_rank} \
    --lora_alpha ${lora_alpha} \
    --trainable ${lora_trainable} \
    --lora_dropout ${lora_dropout} \
    --torch_dtype float16 \
    --validation_file ${validation_file} \
    --ddp_find_unused_parameters False

```

### 依赖情况（代码类问题务必提供）

```
```

### 运行日志或截图

Input:你的作者是谁
Response:  我的作者是OpenAI，一个由一群科学家和工程师组成的人工智能研究团队。他们致力于开发先进的人工智能技术，并为全球用户提供高质量的人工智能服务。

Input:你的名字是什么
Response:  我的名字是Open Assistant，一个开源的聊天机器人项目。
Input:你的身份是什么
Response:  我是Open Assistant，一个由志愿者开发和维护的开源聊天机器人项目。我是一种人工智能技术，旨在通过自然语言处理来回答问题、提供信息或进行对话。我的目标是帮助人们解决问题并提供有用的信息。

**Comment by luhairong11** [comment]
我也遇到同样的情况，用了5W条数据微调，效果不好，也不会回答里面的内容

**Comment by zzx16045223** [comment]
> 我也遇到同样的情况，用了5W条数据微调，效果不好，也不会回答里面的内容
不用那么多数据微调吧，我一千条数据效果都可以，就是会出现遗忘问题

**Comment by zzx16045223** [comment]
> ### 提交前必须检查以下项目
> * [x]  请确保使用的是仓库最新代码（git pull），一些问题已被解决和修复。
> * [x]  我已阅读[项目文档](https://github.com/ymcui/Chinese-LLaMA-Alpaca-2/wiki)和[FAQ章节](https://github.com/ymcui/Chinese-LLaMA-Alpaca-2/wiki/%E5%B8%B8%E8%A7%81%E9%97%AE%E9%A2%98)并且已在Issue中对问题进行了搜索，没有找到相似问题和解决方案。
> * [x]  第三方插件问题：例如[llama.cpp](https://github.com/ggerganov/llama.cpp)、[LangChain](https://github.com/hwchase17/langchain)、[text-generation-webui](https://github.com/oobabooga/text-generation-webui)等，同时建议到对应的项目中查找解决方案。
> 
> ### 问题类型
> 模型训练与精调
> 
> ### 基础模型
> Alpaca-2-7B
> 
> ### 操作系统
> Linux
> 
> ### 详细描述问题
> 对模型进行精调-在完整模型Chinese-LLaMA-Alpaca-2上进行微调， 模型微调运行成功，但是推理的时候代码还是回答的以前问题。请问是哪里出现了问题？我尝试过Chinese-LLaMA-Alpaca-2，以及Chinese-LLaMA-Alpaca 数据集为几十条身份信息数据。 如： { "instruction": "你身份是什么？", "input": "", "output": "我叫ai助手。我的主要目标是协助用户解决问题和满足他们的需求。" }, 而我推理代码如下： `python3 scripts/inference/inference_hf.py \ --base_model /root/models/chinese-alpaca-2-7b \ --lora_model /root/Chinese-LLaMA-Alpaca-2/models/sft_lora_model \ --with_prompt \ --interactive`
> 
> ```
> 微调代码如下
> ########参数部分########
> lr=1e-4
> lora_rank=64
> lora_alpha=128
> lora_trainable="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj"
> lora_dropout=0.05
> 
> pretrained_model=/root/models/chinese-alpaca-2-7b
> chinese_tokenizer_path=/root/models/chinese-alpaca-2-7b
> dataset_dir=/root/Chinese-LLaMA-Alpaca-2/datas/sft
> per_device_train_batch_size=1
> per_device_eval_batch_size=1
> gradient_accumulation_steps=1
> output_dir=/root/Chinese-LLaMA-Alpaca-2/models
> validation_file=/root/Chinese-LLaMA-Alpaca-2/datas/sft/self_cognition.json
> max_seq_length=128
> 
> deepspeed_config_file=ds_zero2_no_offload.json
> 
> ########启动命令########
> torchrun --nnodes 1 --nproc_per_node 2 run_clm_sft_with_peft.py \
>     --deepspeed ${deepspeed_config_file} \
>     --model_name_or_path ${pretrained_model} \
>     --tokenizer_name_or_path ${chinese_tokenizer_path} \
>     --dataset_dir ${dataset_dir} \
>     --per_device_train_batch_size ${per_device_train_batch_size} \
>     --per_device_eval_batch_size ${per_device_eval_batch_size} \
>     --do_train \
>     --do_eval \
>     --seed $RANDOM \
>     --fp16 \
>     --num_train_epochs 2 \
>     --lr_scheduler_type cosine \
>     --learning_rate ${lr} \
>     --warmup_ratio 0.03 \
>     --weight_decay 0 \
>     --logging_strategy steps \
>     --logging_steps 10 \
>     --save_strategy steps \
>     --save_total_limit 3 \
>     --evaluation_strategy steps \
>     --eval_steps 250 \
>     --save_steps 500 \
>     --gradient_accumulation_steps ${gradient_accumulation_steps} \
>     --preprocessing_num_workers 8 \
>     --max_seq_length ${max_seq_length} \
>     --output_dir ${output_dir} \
>     --overwrite_output_dir \
>     --ddp_timeout 30000 \
>     --logging_first_step True \
>     --lora_rank ${lora_rank} \
>     --lora_alpha ${lora_alpha} \
>     --trainable ${lora_trainable} \
>     --lora_dropout ${lora_dropout} \
>     --torch_dtype float16 \
>     --validation_file ${validation_file} \
>     --ddp_find_unused_parameters False
> ```
> 
> ### 依赖情况（代码类问题务必提供）
> ### 运行日志或截图
> Input:你的作者是谁 Response: 我的作者是OpenAI，一个由一群科学家和工程师组成的人工智能研究团队。他们致力于开发先进的人工智能技术，并为全球用户提供高质量的人工智能服务。
> 
> Input:你的名字是什么 Response: 我的名字是Open Assistant，一个开源的聊天机器人项目。 Input:你的身份是什么 Response: 我是Open Assistant，一个由志愿者开发和维护的开源聊天机器人项目。我是一种人工智能技术，旨在通过自然语言处理来回答问题、提供信息或进行对话。我的目标是帮助人们解决问题并提供有用的信息。

请问你微调的时候会出现eval_loss和perplexity为nan吗？

**Comment by luhairong11** [comment]
> > 我也遇到同样的情况，用了5W条数据微调，效果不好，也不会回答里面的内容
> > 不用那么多数据微调吧，我一千条数据效果都可以，就是会出现遗忘问题

方便给一下你的训练参数吗，我用的是作者的默认参数

**Comment by ibmxiang** [comment]
> 请问你微调的时候会出现eval_loss和perplexity为nan吗？
没有发现，不过好像遇到了数字溢出 overflow
'loss': 4.6488, 'learning_rate': 0.0001, 'epoch': 0.09}                                                                             
  5%|████▍                                                                                            | 1/22 [00:00<00:18,  1.13it/s][2023-08-28 10:09:45,138] [INFO] [loss_scaler.py:190:update_scale] [deepspeed] OVERFLOW! Rank 0 Skipping step. Attempted loss scale: 65536, but hysteresis is 2. Reducing hysteresis to 1
 14%|█████████████▏                                                                                   | 3/22 [00:01<00:07,  2.59it/s][2023-08-28 10:09:45,617] [INFO] [loss_scaler.py:183:update_scale] [deepspeed] OVERFLOW! Rank 0 Skipping step. Attempted loss scale: 65536, reducing to 32768
{'loss': 3.382, 'learning_rate': 7.500000000000001e-05, 'epoch': 0.91}                                                               
{'loss': 1.4202, 'learning_rate': 8.688061284200266e-06, 'epoch': 1.82}                                                              
100%|███████████████████████████████████████████████████████████

**Comment by zzx16045223** [comment]
> > 请问你微调的时候会出现eval_loss和perplexity为nan吗？
> > 没有发现，不过好像遇到了数字溢出 overflow
> > 'loss': 4.6488, 'learning_rate': 0.0001, 'epoch': 0.09}
> > 5%|████▍                                                                                            | 1/22 [00:00<00:18,  1.13it/s][2023-08-28 10:09:45,138] [INFO] [loss_scaler.py:190:update_scale] [deepspeed] OVERFLOW! Rank 0 Skipping step. Attempted loss scale: 65536, but hysteresis is 2. Reducing hysteresis to 1
> > 14%|█████████████▏                                                                                   | 3/22 [00:01<00:07,  2.59it/s][2023-08-28 10:09:45,617] [INFO] [loss_scaler.py:183:update_scale] [deepspeed] OVERFLOW! Rank 0 Skipping step. Attempted loss scale: 65536, reducing to 32768
> > {'loss': 3.382, 'learning_rate': 7.500000000000001e-05, 'epoch': 0.91}
> > {'loss': 1.4202, 'learning_rate': 8.688061284200266e-06, 'epoch': 1.82}
> > 100%|███████████████████████████████████████████████████████████
我没有溢出这个问题，如果你没有出现nan的情况，可以把训练相关参数给我参考一下吗？感谢感谢

**Comment by zzx16045223** [comment]
> 

lr=1e-4
lora_rank=64
lora_alpha=128
lora_trainable="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj"
modules_to_save="embed_tokens,lm_head"
lora_dropout=0.05

pretrained_model=/workspace/ypg/Chinese-LLaMA-Alpaca-2/llama2_model/chinese-alpaca-2-13b
chinese_tokenizer_path=/workspace/ypg/Chinese-LLaMA-Alpaca-2/llama2_model/chinese-alpaca-2-13b
dataset_dir=/workspace/ypg/Chinese-LLaMA-Alpaca-2/dataset/cailian_data/20230824_1000_data
per_device_train_batch_size=1
per_device_eval_batch_size=1
gradient_accumulation_steps=8
output_dir=/workspace/ypg/Chinese-LLaMA-Alpaca-2/output/20230824_1000_data/2_epoch
# peft_model=/workspace/ypg/Chinese-LLaMA-Alpaca-2/llama2_model/chinese-alpaca-2-13b
validation_file=/workspace/ypg/Chinese-LLaMA-Alpaca-2/dataset/cailian_data/20230824_1000_data/output.json

num_train_epochs=2
max_seq_length=5000


deepspeed_config_file=ds_zero2_no_offload.json

CUDA_VISIBLE_DEVICES=1 torchrun --nnodes 1 --nproc_per_node 1 run_clm_sft_with_peft.py \
    --deepspeed ${deepspeed_config_file} \
    --model_name_or_path ${pretrained_model} \
    --tokenizer_name_or_path ${chinese_tokenizer_path} \
    --dataset_dir ${dataset_dir} \
    --validation_split_percentage 0.001 \
    --per_device_train_batch_size ${per_device_train_batch_size} \
    --per_device_eval_batch_size ${per_device_eval_batch_size} \
    --do_train \
    --do_eval \
    --seed $RANDOM \
    --fp16 \
    --num_train_epochs ${num_train_epochs} \
    --lr_scheduler_type cosine \
    --learning_rate ${lr} \
    --warmup_ratio 0.03 \
    --weight_decay 0 \
    --logging_strategy steps \
    --logging_steps 10 \
    --save_strategy steps \
    --save_total_limit 3 \
    --evaluation_strategy steps \
    --eval_steps 1000 \
    --save_steps 1000 \
    --gradient_accumulation_steps ${gradient_accumulation_steps} \
    --preprocessing_num_workers 8 \
    --max_seq_length ${max_seq_length} \
    --output_dir ${output_dir} \
    --overwrite_output_dir \
    --ddp_timeout 30000 \
    --logging_first_step True \
    --lora_rank ${lora_rank} \
    --lora_alpha ${lora_alpha} \
    --trainable ${lora_trainable} \
    --modules_to_save ${modules_to_save} \
    --lora_dropout ${lora_dropout} \
    --torch_dtype float16 \
    --validation_file ${validation_file} \
    --gradient_checkpointing \
    --ddp_find_unused_parameters False \
    --flash_attn

    # --peft_path ${peft_model} \

**Comment by zzx16045223** [comment]
> 

你好，这是我的训练脚本
lr=1e-4
lora_rank=64
lora_alpha=128
lora_trainable="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj"
modules_to_save="embed_tokens,lm_head"
lora_dropout=0.05

pretrained_model=/workspace/ypg/Chinese-LLaMA-Alpaca-2/llama2_model/chinese-alpaca-2-13b
chinese_tokenizer_path=/workspace/ypg/Chinese-LLaMA-Alpaca-2/llama2_model/chinese-alpaca-2-13b
dataset_dir=/workspace/ypg/Chinese-LLaMA-Alpaca-2/dataset/cailian_data/20230824_1000_data
per_device_train_batch_size=1
per_device_eval_batch_size=1
gradient_accumulation_steps=8
output_dir=/workspace/ypg/Chinese-LLaMA-Alpaca-2/output/20230824_1000_data/2_epoch
# peft_model=/workspace/ypg/Chinese-LLaMA-Alpaca-2/llama2_model/chinese-alpaca-2-13b
validation_file=/workspace/ypg/Chinese-LLaMA-Alpaca-2/dataset/cailian_data/20230824_1000_data/output.json

num_train_epochs=2
max_seq_length=5000


deepspeed_config_file=ds_zero2_no_offload.json

CUDA_VISIBLE_DEVICES=1 torchrun --nnodes 1 --nproc_per_node 1 run_clm_sft_with_peft.py \
    --deepspeed ${deepspeed_config_file} \
    --model_name_or_path ${pretrained_model} \
    --tokenizer_name_or_path ${chinese_tokenizer_path} \
    --dataset_dir ${dataset_dir} \
    --validation_split_percentage 0.001 \
    --per_device_train_batch_size ${per_device_train_batch_size} \
    --per_device_eval_batch_size ${per_device_eval_batch_size} \
    --do_train \
    --do_eval \
    --seed $RANDOM \
    --fp16 \
    --num_train_epochs ${num_train_epochs} \
    --lr_scheduler_type cosine \
    --learning_rate ${lr} \
    --warmup_ratio 0.03 \
    --weight_decay 0 \
    --logging_strategy steps \
    --logging_steps 10 \
    --save_strategy steps \
    --save_total_limit 3 \
    --evaluation_strategy steps \
    --eval_steps 1000 \
    --save_steps 1000 \
    --gradient_accumulation_steps ${gradient_accumulation_steps} \
    --preprocessing_num_workers 8 \
    --max_seq_length ${max_seq_length} \
    --output_dir ${output_dir} \
    --overwrite_output_dir \
    --ddp_timeout 30000 \
    --logging_first_step True \
    --lora_rank ${lora_rank} \
    --lora_alpha ${lora_alpha} \
    --trainable ${lora_trainable} \
    --modules_to_save ${modules_to_save} \
    --lora_dropout ${lora_dropout} \
    --torch_dtype float16 \
    --validation_file ${validation_file} \
    --gradient_checkpointing \
    --ddp_find_unused_parameters False \
    --flash_attn

**Comment by luxinglong** [comment]
遇到了同样的问题，微调前后推理一样

**Comment by lychee-2724540853** [comment]
之前也遇到了，保存lora模型的时候有问题，保存下来的模型大小有问题

**Comment by xxllp** [comment]
是不是step太小了，训练后的模型没保存呢？
